Add bmp581 driver

talhaHavadar opened the pull request:
Hello all,

Here you can find the implementation of bmp581 driver with I2C for ZephyrRTOS. I am using this driver currently in one of my projects so I am able to test it if something goes wrong. 

Can you please review the changes and give feedback about things I need to change?

I also have question regarding bmp581_user.h Is there a way to open this file to user? Like they can include it in their application code and use some of the definitions in there? I am trying to find best practices to use custom ATTR for example.

Thanks in advance

talhaHavadar committed:
Add support for bmp581 ic

This commit adds source and header files required for bmp581 I2C driver.
I have used bmp581_user.h to add more usage related definitions
but bmp581.h to add hardware related definitions.

Signed-off-by: Talha Can Havadar <havadartalha@gmail.com>
Signed-off-by: Talha Can Havadar <talhahavadar@hotmail.com>

MaureenHelm commented on drivers/sensor/bmp581/bmp581.c:
```diff
@@ -0,0 +1,563 @@
+#ifdef CONFIG_BMP581_TRIGGER
+#define BMP581_CONFIG(i)                                                                           \
+	static const struct bmp581_config bmp581_config_##i = {                                    \
+		.i2c = DEVICE_DT_GET(DT_INST_BUS(i)),                                              \
+		.i2c_addr = DT_INST_REG_ADDR(i),                                                   \
```
Use `I2C_DT_SPEC_INST_GET`

talhaHavadar replied:
Updated as suggested.

MaureenHelm commented on drivers/sensor/bmp581/bmp581.h:
```diff
@@ -0,0 +1,352 @@
+};
+
+#ifdef CONFIG_BMP581_TRIGGER
+int bmp581_trigger_init(const struct device *dev);
+int bmp581_trigger_set(const struct device *dev, const struct sensor_trigger *trig,
+		       sensor_trigger_handler_t handler);
```
Trigger support doesn't appear to be implemented yet. Please remove all trigger references until this feature is actually imoplemented

talhaHavadar replied:
Indeed, I have just removed the not implemented parts from the source.

MaureenHelm commented on dts/bindings/sensor/bosch,bmp581.yml:
```diff
@@ -0,0 +1,12 @@
+
+compatible: "bosch,bmp581"
+
+include: i2c-device.yaml
```
```suggestion
include: [sensor-device.yaml, i2c-device.yaml]
```

MaureenHelm commented on drivers/sensor/bmp581/bmp581.c:
```diff
@@ -0,0 +1,563 @@
+	static struct bmp581_data bmp581_data_##i;                                                 \
+	BMP581_CONFIG(i);                                                                          \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(i, bmp581_init, NULL, &bmp581_data_##i, &bmp581_config_##i,          \
```
Use `SENSOR_DEVICE_DT_INST_DEFINE`

talhaHavadar replied:
Updated as suggested

MaureenHelm reviewed (changes requested):
Please add this sensor to `tests/drivers/build_all/sensor`

> I also have question regarding bmp581_user.h Is there a way to open this file to user? Like they can include it in their application code and use some of the definitions in there? I am trying to find best practices to use custom ATTR for example.

Public definitions can be placed in `include/zephyr/drivers/sensor`

talhaHavadar commented on drivers/sensor/bmp581/Kconfig:
```diff
@@ -0,0 +1,11 @@
+
+menuconfig BMP581
+	bool "BMP581"
+	depends on I2C && GPIO
```
Updated as suggested.

talhaHavadar commented on drivers/sensor/bmp581/bmp581.c:
```diff
@@ -0,0 +1,563 @@
+
+int reg_read(uint8_t reg, uint8_t *data, uint16_t length, struct bmp581_data *drv)
+{
+	return i2c_burst_read(drv->i2c, drv->i2c_addr, reg, data, length);
```
Updated as suggested.

talhaHavadar commented on drivers/sensor/bmp581/bmp581.c:
```diff
@@ -0,0 +1,563 @@
+	case SENSOR_CHAN_ALTITUDE: {
+		// val2 must have P0 (typically pressure at sea level or ground)
+		// calculate altitude in meters
+		float altitude = 44330 * (1.0 - pow(drv->last_sample.pressure / val->val2, 0.1903));
```
I did update the calculation and tried to simplify it a bit more. I also added the link to whitepaper that has the formula.

MaureenHelm reviewed (commented):
Please incorporate requested changes by amending your original commits rather than adding new fixup commits. You also need to address the failed compliance check.

talhaHavadar commented:
> Please incorporate requested changes by amending your original commits rather than adding new fixup commits. You also need to address the failed compliance check.

Just updated the commits and also fixed the coding style issues hopefully. Let's see if it is all green or not.

carlescufi commented:
@gmarull @avisconti could you take a look?

MaureenHelm commented on drivers/sensor/bmp581/bmp581.c:
```diff
@@ -0,0 +1,565 @@
+static int get_power_mode(enum bmp5_powermode *powermode, struct bmp581_data *drv);
+static int set_power_mode(enum bmp5_powermode powermode, struct bmp581_data *drv);
+
+int reg_read(uint8_t reg, uint8_t *data, uint16_t length, struct bmp581_data *drv)
+{
+	return i2c_burst_read_dt(&drv->i2c, reg, data, length);
+}
+
+int reg_write(uint8_t reg, const uint8_t *data, uint16_t length, struct bmp581_data *drv)
+{
+	return i2c_burst_write_dt(&drv->i2c, reg, data, length);
+}
```
Does this sensor also support SPI? If not, I don't see value in this wrapper layer.

talhaHavadar replied:
yep sensor also supports SPI. But for now I didnt have time to implement SPI so I put it this way to make it easy to add in future.

MaureenHelm commented on drivers/sensor/bmp581/bmp581.c:
```diff
@@ -0,0 +1,565 @@
+	memset(&drv->osr_odr_press_config, 0, sizeof(drv->osr_odr_press_config));
+	memset(&drv->last_sample, 0, sizeof(drv->last_sample));
+
+	drv->i2c = cfg->i2c;
+
+	drv->i2c_addr = cfg->i2c_addr;
```
Why is this necessary?

talhaHavadar replied:
It is probably legacy. Let me check in next commit

MaureenHelm commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -483,8 +483,8 @@ test_i2c_mpr: mpr@18 {
 test_i2c_dps310: dps310@48 {
-		compatible = "infineon,dps310";
-		reg = <0x48>;
+	compatible = "infineon,dps310";
+	reg = <0x48>;
```
This change is unrelated to the bmp581 driver. Please revert or fix in a separate commit

MaureenHelm reviewed (changes requested):
> You also need to address the failed compliance check.

This still needs to be fixed.
